packages/pyright-internal/src/languageService/definitionProvider.ts

```diff
@@ -207,6 +219,65 @@ export class DefinitionProvider {
         }
     }
 
+    private static _tryAddTypedDictDefinitions(
+        node: StringNode,
+        evaluator: TypeEvaluator,
+        definitions: DocumentRange[],
+        sourceMapper: SourceMapper
+    ) {
+        const type = evaluator.getExpectedType(node)?.type;
+        if (type === undefined) {
+            return;
+        }
+
+        doForEachSubtype(type, (subtype) => {
+            if (isClassInstance(subtype) && ClassType.isTypedDictClass(subtype)) {
+                const entry = subtype.details.typedDictEntries?.get(node.value);
+                if (entry !== undefined) {
+                    DefinitionProvider._resolveDeclarations(
+                        DefinitionProvider._resolveField(subtype, node.value)?.getDeclarations(),
+                        evaluator,
+                        definitions,
+                        sourceMapper
+                    );
+                }
+            }
+        });
+    }
+
+    private static _resolveField(type: ClassType, name: string, checkedTypes?: Set<TypeSourceId>): Symbol | undefined {
+        // resolve a field from the symbol table for the given class and all it's base classes
+        let field = type.details.fields.get(name);
+        if (field !== undefined) {
+            return field;
+        }
+
+        // recursion guard
+        if (checkedTypes === undefined) {
+            checkedTypes = new Set();
+        } else if (checkedTypes.has(type.details.typeSourceId)) {
+            // returning undefined is fine as if we've checked this type before
+            // then it didn't include the field as otherwise we would have already returned
+            return undefined;
+        }
+
+        checkedTypes.add(type.details.typeSourceId);
+
+        for (let index = 0; index < type.details.baseClasses.length; index++) {
+            const base = type.details.baseClasses[index];
+            if (!isClass(base)) {
+                continue;
+            }
+
+            field = DefinitionProvider._resolveField(base, name, checkedTypes);
+            if (field !== undefined) {
+                return field;
+            }
+        }
+
+        return undefined;
+    }
```

Comment:
I'm not sure if this is the correct place for this function. I was also overly cautious with a recursion guard that may not be necessary.

I also would be surprised if a utility function like this didn't already exist somewhere in the codebase but I couldn't find it. If such a function does exist I'd be more than happy to refactor my code to use that instead.

Reply:
This function should probably be in the type evaluator, not in the definition provider. It's analogous to `getDeclarationsForNameNode` but for a StringNode, so maybe call it `getDeclarationsForStringNode`.

Within the type evaluator, you can use `lookUpObjectMember` to find the declaration. This will use proper MRO ordering and avoid returning overridden declarations. Refer to the memberAccess section of `getDeclarationsForNameNode`.